arangod/Aql/Functions.cpp

```diff
@@ -4569,6 +4606,158 @@ AqlValue Functions::RandomToken(ExpressionContext*, transaction::Methods*,
+
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  if (!value.isString()) {
+    registerWarning(expressionContext, AFN, TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH);
+    return AqlValue(AqlValueHintNull());
+  }
+
+  int64_t result = 0;
+
+  AqlValueMaterializer materializer(trx);
+  VPackSlice slice = materializer.slice(value, false);
+
+  // parse the input string
+  TRI_ASSERT(slice.isString());
+  VPackValueLength l;
+  char const* p = slice.getString(l);
+  char const* e = p + l;
+
+  char const* current = p;
+  constexpr size_t parts = 4;
+  for (size_t i = 0; i < parts; ++i) {
+    if (i > 0) {
+      if (current < e && *current == '.') {
+        ++current;
+      } else {
+        registerInvalidArgumentWarning(expressionContext, AFN);
+        return AqlValue(AqlValueHintNull());
+      }
+    }
+    char const* start = current;
+    while (current < e && *current >= '0' && *current <= '9') {
+      ++current;
+    }
+
+    // max expected value is 3 digits long
+    if ((current == start) || (current - start) > 3) {
+      registerInvalidArgumentWarning(expressionContext, AFN);
+      return AqlValue(AqlValueHintNull());
+    }
+    int64_t number = NumberUtils::atoi_unchecked<int64_t>(start, current);
+    if (number > 255) {
+      registerInvalidArgumentWarning(expressionContext, AFN);
+      return AqlValue(AqlValueHintNull());
+    }
+    result += number << ((parts - 1 - i) * 8);
+  }
+
+  if (current != e) {
+    // trailing data
+    registerInvalidArgumentWarning(expressionContext, AFN);
+    return AqlValue(AqlValueHintNull());
+  }
+
+  return AqlValue(AqlValueHintUInt(result));
+}
+
+/// @brief function IS_IP4
+AqlValue Functions::IsIp4(ExpressionContext* expressionContext, transaction::Methods* trx,
+                          VPackFunctionParameters const& parameters) {
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  if (!value.isString()) {
+    return AqlValue(AqlValueHintBool(false));
+  }
+
+  AqlValueMaterializer materializer(trx);
+  VPackSlice slice = materializer.slice(value, false);
+
+  // parse the input string
+  TRI_ASSERT(slice.isString());
+  VPackValueLength l;
+  char const* p = slice.getString(l);
+  char const* e = p + l;
+
+  char const* current = p;
+  constexpr size_t parts = 4;
+  for (size_t i = 0; i < parts; ++i) {
+    if (i > 0) {
+      if (current < e && *current == '.') {
+        ++current;
+      } else {
+        return AqlValue(AqlValueHintBool(false));
+      }
+    }
+    char const* start = current;
+    while (current < e && *current >= '0' && *current <= '9') {
+      ++current;
+    }
+
+    // max expected value is 3 digits long
+    bool valid = ((current != start) && (current - start) <= 3);
+    if (valid) {
+      int64_t number = NumberUtils::atoi_unchecked<int64_t>(start, current);
+      valid = (number <= 255);
+    }
+    if (!valid) {
+      return AqlValue(AqlValueHintBool(false));
+    }
+  }
```

Comment:
Same here, can't we call e.g. `inet_pton` and check the return value?

Reply:
Done